src/App.ts

```diff
@@ -0,0 +1,117 @@
+/**
+ * Italia PagoPA Proxy
+ * Cittadinanza Digitale PagoPA services
+ *
+ */
+
+import * as bodyParser from "body-parser";
+import * as express from "express";
+import * as morgan from "morgan";
+
+import { none, Option, some } from "fp-ts/lib/Option";
+
+import { UserController } from "./controllers/UserController";
+import { CreditCard } from "./types/CreditCard";
+import { PaymentMethod, PaymentMethodType } from "./types/PaymentMethod";
+import { IUser } from "./types/User";
+
+const VERSION: string = "0.0.11";
+const NAME: string = "italia-pagopa-proxy";
+
+function getUserAndThen<T>(
+    req: express.Request,
+    res: express.Response,
+    andThen: ((userController: UserController) => Option<T>)
+): Option<T> {
+    try {
+        const token = req.query.token;
+        const user: IUser = { token };
+        const userController = new UserController(user);
+        return andThen(userController);
+    } catch (err) {
+        // TODO: define what status code to return
+        res.json({
+            status: "ERROR",
+            message: err.message
+        });
+    }
+    return none;
+}
+
+export function newApp(
+    /* TODO: receive env to decide whether to enable ssl */
+    testMode: boolean = false
+): express.Express {
+    const app = express(); // create app
+
+    // setup Express app
+    if (!testMode) {
+        app.use(morgan("combined"));
+    }
+    app.use(bodyParser.json());
+    app.use(bodyParser.urlencoded({ extended: false }));
+
+    // setup routes
+    app.get("/", (_, res) => {
+        res.json({
+            message: `Welcome to ${NAME}!`,
+            version: VERSION
+        });
+    });
+
+    app.get("/getCreditCards", (_, res) => {
+        res.json();
+    });
+
+    app.get("/wallet", (req, res) => {
+        const wallet: Option<ReadonlyArray<PaymentMethod>> = getUserAndThen<
+            ReadonlyArray<PaymentMethod>
+        >(req, res, (user: UserController) => some(user.getWallet()));
```

Comment:
this should be in the form

`getUserFromRequest(req).fold(
  () => [],
  user => user.getWallet()
)`

where getUserFromRequest (sync) returns a `Option<User>` and probably `user.getWallet` should return an `Either<Error, Option<User>>` (I'm not aware of the getWallet internals)

the general rules are:
- if the function may return null or undefined, return an `Option<BusinessObject>` instead
- if the function may return an error return an `Either<Error, BusinessObject>`
- if the function may return an Error or a BusinessObject or null / undefined return` Either<Error, Option<BO>>`
- if the function is async return a Promise that wraps the Option or the Either, ie. 
 Promise<Either<Error, Option<BO>>